### DISCUSSION TITLE
skywalking log monitoring

### DISCUSSION THREAD

**Discussion by niharranjannihar** [apache/skywalking#12026] [Q&A]
Dear Team,
Does anyone have any idea how to monitor logs of services in skywalking .
![Screenshot from 2024-03-15 17-27-46](https://github.com/apache/skywalking/assets/106923369/4bdbbf58-3bba-4b6a-a368-0a005606ed4f)

**Comment by wu-sheng** [comment]
You could simply ask this on website,ASK AI widget.
It will show you full list.

**Comment by wu-sheng** [comment]
# Copied from ASK AI

SkyWalking provides the capability to monitor logs of services. Here are the steps to set it up:

1. **Setup Log Collection**: SkyWalking provides log collection and error log analysis. You can use Fluent-bit to collect and report access logs and error logs to SkyWalking for analysis and storage. The Fluent-bit configuration defines the log collection directory and the logs will be reported through the rest port of OAP after being processed by specific functions. Here is an example of the Fluent-bit configuration:

```plaintext
[SERVICE]
    Flush          5
    Daemon         Off
    Log_Level      warn
[INPUT]
    Name           tail
    Tag            access
    Path           /var/log/nginx/access.log
[INPUT]
    Name           tail
    Tag            error
    Path           /var/log/nginx/error.log
[FILTER]
    Name           lua
    Match          access
    Script         fluent-bit-script.lua
    Call           rewrite_access_log
[FILTER]
    Name           lua
    Match          error
    Script         fluent-bit-script.lua
    Call           rewrite_error_log
[OUTPUT]
    Name            stdout
    Match           *
    Format          json
[OUTPUT]
    Name            http
    Match           *
    Host            oap
    Port            12800
    URI             /v3/logs
    Format          json
```
(Source: [Monitoring Nginx with SkyWalking](https://skywalking.apache.org/blog/2023-12-23-monitoring-nginx-by-skywalking#access--error-log-collection))

2. **View Logs on UI**: After setting up Fluent-bit, you can view the collected logs in the Log tab of the SkyWalking monitoring dashboard.
(Source: [使用 SkyWalking 监控 Nginx](https://skywalking.apache.org/zh/2023-12-23-monitoring-nginx-by-skywalking#access--error-log-%E9%87%87%E9%9B%86))

Please note that the specific setup might vary depending on your service and infrastructure.

**Comment by wu-sheng** [comment]
If you want to OTEL log format, copied from ASK AI again

Yes, you can use OpenTelemetry (OTel) to send log data to SkyWalking. Here are the steps to set it up:

1. **Enable OTel Receiver**: Make sure the `receiver-otel` is enabled in your SkyWalking configuration. This can be done in the `application.yml` file. Here is an example:

```yaml
receiver-otel:
 selector: ${SW_OTEL_RECEIVER:default}
 default:
 enabledHandlers: ${SW_OTEL_RECEIVER_ENABLED_HANDLERS:"otlp-metrics,otlp-logs"}
```
(Source: [OpenTelemetry Logging Format > OpenTelemetry OTLP Exporter](https://skywalking.apache.org/docs/main/latest/en/setup/backend/log-otlp#opentelemetry-otlp-exporter))

2. **Ensure Log Data Contains Required Attributes**: Most of the language SDKs of OpenTelemetry do not support logging feature (yet) or the logging feature is experimental. It’s your responsibility to make sure the reported log data contains the following attributes, otherwise SkyWalking is not able to consume them:
   - `service.name`: the name of the service that generates the log data. OpenTelemetry Java SDK (experimental) has this attribute set. If you’re using other SDK or agent, please check the corresponding documentation.
   (Source: [OpenTelemetry Logging Format > OpenTelemetry OTLP Exporter](https://skywalking.apache.org/docs/main/latest/en/setup/backend/log-otlp#opentelemetry-otlp-exporter))

Please note that the specific setup might vary depending on your service and infrastructure.

**Comment by niharranjannihar** [comment]
root@administrator:/etc/fluent-bit# cat fluent-bit.conf 
[SERVICE]
    # Flush
    # =====
    # set an interval of seconds before to flush records to a destination
    flush        1

    # Daemon
    # ======
    # instruct Fluent Bit to run in foreground or background mode.
    daemon       Off

    # Log_Level
    # =========
    # Set the verbosity level of the service, values can be:
    #
    # - error
    # - warning
    # - info
    # - debug
    # - trace
    #
    # by default 'info' is set, that means it includes 'error' and 'warning'.
    log_level    info

    # Parsers File
    # ============
    # specify an optional 'Parsers' configuration file
    #parser_file fluent-bit-parser.conf

    # Plugins File
    # ============
    # specify an optional 'Plugins' configuration file to load external plugins.
    #plugins_file plugins.conf

    # HTTP Server
    # ===========
    # Enable/Disable the built-in HTTP Server for metrics
    http_server  Off
    http_listen  0.0.0.0
    http_port    9091

    # Storage
    # =======
    # Fluent Bit can use memory and filesystem buffering based mechanisms
    #
    # - https://docs.fluentbit.io/manual/administration/buffering-and-storage
    #
    # storage metrics
    # ---------------
    # publish storage pipeline metrics in '/api/v1/storage'. The metrics are
    # exported only if the 'http_server' option is enabled.
    #
    #storage.metrics on

    # storage.path
    # ------------
    # absolute file system path to store filesystem data buffers (chunks).
    #
    # storage.path /tmp/storage

    # storage.sync
    # ------------
    # configure the synchronization mode used to store the data into the
    # filesystem. It can take the values normal or full.
    #
    # storage.sync normal

    # storage.checksum
    # ----------------
    # enable the data integrity check when writing and reading data from the
    # filesystem. The storage layer uses the CRC32 algorithm.
    #
    # storage.checksum off

    # storage.backlog.mem_limit
    # -------------------------
    # if storage.path is set, Fluent Bit will look for data chunks that were
    # not delivered and are still in the storage layer, these are called
    # backlog data. This option configure a hint of maximum value of memory
    # to use when processing these records.
    #
    # storage.backlog.mem_limit 5M

[INPUT]
    name tail
    path ***************************************************

[OUTPUT]
    name  stdout
    match *
[FILTER]
    name  lua
    match *
    script fluent-bit-script.lua
    call  rewrite_body
[OUTPUT]
    name  http
    match *
    host  0.0.0.0
    port  12800
    uri   /v3/logs
    format json
    
    
    
root@administrator:/etc/fluent-bit# cat fluent-bit-script.lua 
function rewrite_body(tag, timestamp, record)
    local newRecord = {}
    newRecord["service"] = "nodejs::Tomcat-test"
    newRecord["serviceInstance"] = "APMOSYSLT0724"
    newRecord["body"] = { text = { text = record.log } }
    newRecord["tags"] = { data = {{ key = "level", value = "INFO"}}}
    return 1, timestamp, newRecord
end




root@administrator:/etc/fluent-bit# journalctl -u fluent-bit -f
-- Logs begin at Sat 2024-03-16 14:09:39 IST. --
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [sp] stream processor started
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [output:http:http.1] worker #1 started
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [input:tail:tail.0] inotify_fs_add(): inode=15385333 watch_fd=1 name=***************************************************
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [output:http:http.1] worker #0 started
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [input:tail:tail.0] inotify_fs_add(): inode=15385335 watch_fd=2 name=***************************************************
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [input:tail:tail.0] inotify_fs_add(): inode=15402757 watch_fd=3 name=***************************************************
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [input:tail:tail.0] inotify_fs_add(): inode=15384057 watch_fd=4 name=***************************************************
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [input:tail:tail.0] inotify_fs_add(): inode=15402483 watch_fd=5 name=***************************************************
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [input:tail:tail.0] inotify_fs_add(): inode=15385334 watch_fd=6 name=***************************************************
Mar 16 19:36:15 administrator fluent-bit[63346]: [2024/03/16 19:36:15] [ info] [input:tail:tail.0] inotify_fs_add(): inode=15385332 watch_fd=7 name=***************************************************



After doing all the changes still I am unable to see the logs in skywalking UI .

Kindly help me on this .

Regards
Nihar

**Reply by wu-sheng** [reply]
If you want to make it work, you need to write LAL, https://skywalking.apache.org/docs/main/latest/en/concepts-and-designs/lal/

You can find LAL script examples, https://github.com/apache/skywalking/tree/master/oap-server/server-starter/src/main/resources/lal

Make sure you have set up the layer accordingly.

___

I am not sure what your use cases are, if you have Java agent setup, it would be easy to add log4j integration without any manual configurations. https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/application-toolkit-log4j-1.x/

**Reply by wu-sheng** [reply]
I don't have time to debug your script. Sorry. You need to make sure they are following the expectation and LAL script extracts all necessary fields, especially, timebucket, service name, layer. Then you should be able to see services first, then logs if that layer has added logs widgets by default. Otherwise, you need to update UI dashboard.

As I mentioned, General Layer from agent is the easiest way to do, and most users use in that way.
